File: structural-design-pattern/decorator-pattern/decorator.go
// Decorator Pattern
// =================
// Category: Structural Design Pattern
//
// Intent:
//   Attach additional responsibilities to an object dynamically by wrapping it.
//   Each decorator adds one behavior and delegates to the wrapped object.
//
// When to use:
//   - When you want to add behavior without modifying the original type
//   - When behaviors need to be combined in arbitrary ways at runtime
//   - Follows Open/Closed Principle
//
// Key Participants:
//   - Component (interface): Base interface
//   - ConcreteComponent: The original object
//   - Decorator: Wraps the interface and adds behavior

package main

import (
	"fmt"
	"strconv"
	"strings"
	"unicode"
)

// ---------------------------------------------------------------------------
// Component interface
// ---------------------------------------------------------------------------

type DataSource interface {
	Write(data string) string
	Read(data string) string
}

// ---------------------------------------------------------------------------
// Concrete Component
// ---------------------------------------------------------------------------

type StringDataSource struct{}

func (s *StringDataSource) Write(data string) string { return data }
func (s *StringDataSource) Read(data string) string  { return data }

// ---------------------------------------------------------------------------
// Encryption Decorator
// ---------------------------------------------------------------------------

type EncryptionDecorator struct {
	wrapped DataSource
	shift   int
}

func NewEncryptionDecorator(wrapped DataSource, shift int) *EncryptionDecorator {
	return &EncryptionDecorator{wrapped: wrapped, shift: shift}
}

func (e *EncryptionDecorator) shiftChar(c rune, forward bool) rune {
	if c >= 32 && c < 127 {
		var shifted int
		if forward {
			shifted = int(((int(c)-32+e.shift)%95)+32)
		} else {
			shifted = int(((int(c)-32-e.shift+95)%95)+32)
		}
		return rune(shifted)
	}
	return c
}

func (e *EncryptionDecorator) Write(data string) string {
	var encrypted strings.Builder
	for _, c := range data {
		encrypted.WriteRune(e.shiftChar(c, true))
	}
	return e.wrapped.Write(encrypted.String())
}

func (e *EncryptionDecorator) Read(data string) string {
	inner := e.wrapped.Read(data)
	var decrypted strings.Builder
	for _, c := range inner {
		decrypted.WriteRune(e.shiftChar(c, false))
	}
	return decrypted.String()
}

// ---------------------------------------------------------------------------
// Compression Decorator (simple RLE)
// ---------------------------------------------------------------------------

type CompressionDecorator struct {
	wrapped DataSource
}

func NewCompressionDecorator(wrapped DataSource) *CompressionDecorator {
	return &CompressionDecorator{wrapped: wrapped}
}

func rleEncode(data string) string {
	if len(data) == 0 {
		return ""
	}
	runes := []rune(data)
	var result strings.Builder
	count := 1

	for i := 1; i <= len(runes); i++ {
		if i < len(runes) && runes[i] == runes[i-1] {
			count++
		} else {
			if count > 1 {
				result.WriteString(strconv.Itoa(count))
			}
			result.WriteRune(runes[i-1])
			count = 1
		}
	}
	return result.String()
}

func rleDecode(data string) string {
	var result strings.Builder
	num := ""

	for _, c := range data {
		if unicode.IsDigit(c) {
			num += string(c)
		} else {
			count := 1
			if num != "" {
				count, _ = strconv.Atoi(num)
				num = ""
			}
			result.WriteString(strings.Repeat(string(c), count))
		}
	}
	return result.String()
}

func (c *CompressionDecorator) Write(data string) string {
	return c.wrapped.Write(rleEncode(data))
}

func (c *CompressionDecorator) Read(data string) string {
	return rleDecode(c.wrapped.Read(data))
}

// ---------------------------------------------------------------------------
// Logging Decorator
// ---------------------------------------------------------------------------

type LoggingDecorator struct {
	wrapped DataSource
}

func NewLoggingDecorator(wrapped DataSource) *LoggingDecorator {
	return &LoggingDecorator{wrapped: wrapped}
}

func (l *LoggingDecorator) Write(data string) string {
	fmt.Printf("  [LOG] Writing %d chars\n", len(data))
	result := l.wrapped.Write(data)
	fmt.Printf("  [LOG] After processing: %d chars\n", len(result))
	return result
}

func (l *LoggingDecorator) Read(data string) string {
	fmt.Printf("  [LOG] Reading %d chars\n", len(data))
	result := l.wrapped.Read(data)
	fmt.Printf("  [LOG] After processing: %d chars\n", len(result))
	return result
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
func main() {
	fmt.Println("=== Plain write/read ===")
	source := &StringDataSource{}
	written := source.Write("Hello, World!")
	fmt.Println("Written:", written)
	fmt.Println("Read:   ", source.Read(written))

	// Stack: Logging → Encryption → Compression → Base
	fmt.Println("\n=== Decorated: Logging + Encryption + Compression ===")
	decorated := NewLoggingDecorator(
		NewEncryptionDecorator(
			NewCompressionDecorator(&StringDataSource{}),
			5,
		),
	)

	data := "aaabbbccc Hello World!!!"
	processed := decorated.Write(data)
	fmt.Println("Original: ", data)
	fmt.Println("Processed:", processed)

	fmt.Println("\nReading back:")
	restored := decorated.Read(processed)
	fmt.Println("Restored: ", restored)
	fmt.Println("Match:    ", data == restored)
}

File: structural-design-pattern/decorator-pattern/decorator.rs
//! Decorator Pattern
//! =================
//! Category: Structural Design Pattern
//!
//! Intent:
//!   Attach additional responsibilities to an object dynamically by wrapping it.
//!   Each decorator adds one behavior and delegates to the wrapped object.
//!
//! When to use:
//!   - When you want to add behavior without modifying the original type
//!   - When behaviors need to be combined in arbitrary ways at runtime
//!   - Follows Open/Closed Principle
//!
//! Key Participants:
//!   - Component (trait): Base interface
//!   - ConcreteComponent: The original object
//!   - Decorator: Wraps a trait object and adds behavior

// ---------------------------------------------------------------------------
// Component trait
// ---------------------------------------------------------------------------
trait DataSource {
    fn write(&self, data: &str) -> String;
    fn read(&self, data: &str) -> String;
}

// ---------------------------------------------------------------------------
// Concrete Component
// ---------------------------------------------------------------------------
struct StringDataSource;

impl DataSource for StringDataSource {
    fn write(&self, data: &str) -> String {
        data.to_string()
    }
    fn read(&self, data: &str) -> String {
        data.to_string()
    }
}

// ---------------------------------------------------------------------------
// Encryption Decorator
// ---------------------------------------------------------------------------
struct EncryptionDecorator {
    wrapped: Box<dyn DataSource>,
    shift: u8,
}

impl EncryptionDecorator {
    fn new(wrapped: Box<dyn DataSource>, shift: u8) -> Self {
        Self { wrapped, shift }
    }

    fn shift_char(&self, c: char, forward: bool) -> char {
        let code = c as u32;
        if (32..127).contains(&code) {
            let shifted = if forward {
                ((code - 32 + self.shift as u32) % 95) + 32
            } else {
                ((code - 32 + 95 - self.shift as u32) % 95) + 32
            };
            char::from_u32(shifted).unwrap_or(c)
        } else {
            c
        }
    }
}

impl DataSource for EncryptionDecorator {
    fn write(&self, data: &str) -> String {
        let encrypted: String = data.chars().map(|c| self.shift_char(c, true)).collect();
        self.wrapped.write(&encrypted)
    }

    fn read(&self, data: &str) -> String {
        let inner = self.wrapped.read(data);
        inner.chars().map(|c| self.shift_char(c, false)).collect()
    }
}

// ---------------------------------------------------------------------------
// Compression Decorator (simple RLE)
// ---------------------------------------------------------------------------
struct CompressionDecorator {
    wrapped: Box<dyn DataSource>,
}

impl CompressionDecorator {
    fn new(wrapped: Box<dyn DataSource>) -> Self {
        Self { wrapped }
    }

    fn rle_encode(data: &str) -> String {
        if data.is_empty() {
            return String::new();
        }
        let chars: Vec<char> = data.chars().collect();
        let mut result = String::new();
        let mut count = 1;

        for i in 1..=chars.len() {
            if i < chars.len() && chars[i] == chars[i - 1] {
                count += 1;
            } else {
                if count > 1 {
                    result.push_str(&count.to_string());
                }
                result.push(chars[i - 1]);
                count = 1;
            }
        }
        result
    }

    fn rle_decode(data: &str) -> String {
        let mut result = String::new();
        let mut num = String::new();

        for c in data.chars() {
            if c.is_ascii_digit() {
                num.push(c);
            } else {
                let count: usize = num.parse().unwrap_or(1);
                for _ in 0..count {
                    result.push(c);
                }
                num.clear();
            }
        }
        result
    }
}

impl DataSource for CompressionDecorator {
    fn write(&self, data: &str) -> String {
        let compressed = Self::rle_encode(data);
        self.wrapped.write(&compressed)
    }

    fn read(&self, data: &str) -> String {
        let inner = self.wrapped.read(data);
        Self::rle_decode(&inner)
    }
}

// ---------------------------------------------------------------------------
// Logging Decorator
// ---------------------------------------------------------------------------
struct LoggingDecorator {
    wrapped: Box<dyn DataSource>,
}

impl LoggingDecorator {
    fn new(wrapped: Box<dyn DataSource>) -> Self {
        Self { wrapped }
    }
}

impl DataSource for LoggingDecorator {
    fn write(&self, data: &str) -> String {
        println!("  [LOG] Writing {} chars", data.len());
        let result = self.wrapped.write(data);
        println!("  [LOG] After processing: {} chars", result.len());
        result
    }

    fn read(&self, data: &str) -> String {
        println!("  [LOG] Reading {} chars", data.len());
        let result = self.wrapped.read(data);
        println!("  [LOG] After processing: {} chars", result.len());
        result
    }
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
fn main() {
    println!("=== Plain write/read ===");
    let source = StringDataSource;
    let written = source.write("Hello, World!");
    println!("Written: {}", written);
    println!("Read:    {}", source.read(&written));

    // Stack: Logging → Encryption → Compression → Base
    println!("\n=== Decorated: Logging + Encryption + Compression ===");
    let decorated = LoggingDecorator::new(Box::new(EncryptionDecorator::new(
        Box::new(CompressionDecorator::new(Box::new(StringDataSource))),
        5,
    )));

    let data = "aaabbbccc Hello World!!!";
    let processed = decorated.write(data);
    println!("Original:  {}", data);
    println!("Processed: {}", processed);

    println!("\nReading back:");
    let restored = decorated.read(&processed);
    println!("Restored:  {}", restored);
    println!("Match:     {}", data == restored);
}
